src/nodes/flow/flow_models/flow_model.jl

```diff
@@ -0,0 +1,531 @@
+function forward!(output::Vector{T1}, model::CompiledFlowModel, input::Vector{T2}) where { T1 <: Real, T2 <: Real }
+
+    # fetch layers
+    layers = getlayers(model)
+
+    # promote type for allocating output
+    T = promote_type(eltype(model), T2)
+
+    # decouple changing input from actual input
+    input_new = zeros(T, size(input))
+    input_new .= input
+
+    # perform forward pass over all layers
+    forward!(output, layers, input_new)
+    
+end
+
+# inplace forward pass through a tuple of layers
+function forward!(output::Vector{T1}, layers::T, input::Vector{T2}) where { T1 <: Real, T2 <: Real,  T <: NTuple{N,AbstractLayer} where N}
+
+    # perform pass over first layer
+    forward!(output, first(layers), input)
+
+    # update intermediate input
+    input .= output
+
+    # perform pass over all remaining layers
+    forward!(output, Base.tail(layers), input)
+
+end
+
+# when no layers are left, stop the inplace recursion
+forward!(output::Vector{T1}, layers::Tuple{}, input::Vector{T2}) where { T1 <: Real, T2 <: Real } = return
+
+
+# backward pass through the Flow model
+function _backward(model::CompiledFlowModel, output::Vector{T1}) where { T1 <: Real }
+
+    # promote type for allocating output
+    T = promote_type(eltype(model), T1)
+
+    # allocate space for result
+    input = zeros(T, size(output))
+
+    # perform calculations 
+    backward!(input, model, output)
+
+    # return result
+    return input
+    
+end
+
+# when calling backward, redirect to _backward
+backward(model::CompiledFlowModel, output::Vector{T}) where { T <: Real } = _backward(model, output)
+
+# for broadcasting over backward, fix the model for multiple inputs
+Broadcast.broadcasted(::typeof(backward), model::CompiledFlowModel, output::Vector{Vector{T}}) where { T <: Real } = broadcast(_backward, Ref(model), output)
+
+# inplace backward pass through the Flow model
+function backward!(input::Vector{T1}, model::CompiledFlowModel, output::Vector{T2}) where { T1 <: Real, T2 <: Real }
+
+    # fetch layers
+    layers = getlayers(model)
+
+    # promote type for allocating output
+    T = promote_type(eltype(model), T2)
+
+    # decouple changing output from actual output
+    output_new = zeros(T, size(output))
+    output_new .= output
+
+    # perform backward pass over all layers
+    backward!(input, layers, output_new)
+    
+end
+
+# inplace backward pass through a tuple of layers
+function backward!(input::Vector{T1}, layers::T, output::Vector{T2}) where { T1 <: Real, T2 <: Real,  T <: NTuple{N,AbstractLayer} where N}
+
+    # perform pass over last layer
+    backward!(input, last(layers), output)
+
+    # update intermediate output
+    output .= input
+
+    # perform pass over all remaining layers
+    backward!(input, Base.front(layers), output)
+
+end
+
+# when no layers are left, stop the inplace recursion
+backward!(input::Vector{T1}, layers::Tuple{}, output::Vector{T2}) where { T1 <: Real, T2 <: Real } = return
```

Comment:
Just again, better to replace `Vector` with `AbstractVector` everywhere.

Reply:
Resolved by a3da751.

Reply:
Resolved by 0056d3a.